vermouth/processors/do_links.py

```diff
@@ -261,36 +262,147 @@ def _build_link_interaction_from(molecule, interaction, match):
+                                  tuple([new_interaction.meta.get("version", 0)])
+
+                interaction_key_rev = tuple(new_interaction.atoms[::-1]) +\
+                                      tuple([new_interaction.meta.get("version", 0)])
+
+                if interaction_key_rev in self.applied_links[inter_type]:
+                    self.applied_links[inter_type][interaction_key_rev] = (new_interaction,
+                                                                           self.current_link.citations)
+                else:
+                    self.applied_links[inter_type][interaction_key] = (new_interaction,
+                                                                       self.current_link.citations)
+        return molecule
+
+    def remove_interactions(self, molecule, match):
+        """
+        Remove interactions from link to the `self.applied_interactions` dict.
+
+        Parameters
+        ----------
+        molecule: :class:`vermouth.Molecule`
+        match: dict
+            dict relating nodes in `self.current_link` to those in molecule
+        """
+        for inter_type in self.current_link.removed_interactions:
+            for interaction in self.current_link.removed_interactions[inter_type]:
+                new_interaction = _build_link_interaction_from(molecule, interaction, match)
+                interaction_key = tuple(new_interaction.atoms) +\
+                                  tuple([new_interaction.meta.get("version", 0)])
+
+                interaction_key_rev = tuple(new_interaction.atoms[::-1]) +\
+                                      tuple([new_interaction.meta.get("version", 0)])
+
+                if interaction_key in self.applied_links[inter_type]:
+                    del self.applied_links[inter_type][interaction_key]
+
+                if interaction_key_rev in self.applied_links[inter_type]:
+                    del self.applied_links[inter_type][interaction_key]
+        return molecule
+
+    def add_edges(self, molecule, match):
+        """
+        Add edges from link to the molecule.
+
+        Parameters
+        ----------
+        molecule: :class:`vermouth.Molecule`
+        match: dict
+            dict relating nodes in `self.current_link` to those in molecule
+        """
+        # now we already add the edges of this link
+        # links can overwrite each other but the edges must be the same
+        # this is safer than using the make_edge method because it accounts
+        # for edges written in the edges directive
+        for edge in self.current_link.edges:
+            molecule.add_edge(match[edge[0]], match[edge[1]])
+        return molecule
+
     def run_molecule(self, molecule):
-        # TODO: Separate this into a function
         links = molecule.force_field.links
         _nodes_to_remove = []
+        # loop over all links in force-field
         for link in links:
+            self.current_link = link
```

Comment:
Why not pass `link` as argument to the functions/methods?

Reply:
mehh I thought it was a clever choice today but I see the point